client/components/data/query-sites/index.jsx

```diff
@@ -17,66 +13,74 @@ import { requestSites, requestSite } from 'state/sites/actions';
 import { getPreference } from 'state/preferences/selectors';
 import getPrimarySiteId from 'state/selectors/get-primary-site-id';
 
-class QuerySites extends Component {
-	UNSAFE_componentWillMount() {
-		this.requestAll( this.props );
-		this.requestPrimary( this.props );
-		this.requestRecent( this.props );
-		this.requestSingle( this.props );
-	}
-
-	UNSAFE_componentWillReceiveProps( nextProps ) {
-		if ( nextProps.siteId !== this.props.siteId ) {
-			this.requestSingle( nextProps );
-		}
+const getRecentSites = state => getPreference( state, 'recentSites' );
 
-		if ( nextProps.primaryAndRecent && nextProps.primarySiteId !== this.props.primarySiteId ) {
-			this.requestPrimary( nextProps );
-		}
+const requestAll = () => ( dispatch, getState ) => {
+	if ( ! isRequestingSites( getState() ) ) {
+		dispatch( requestSites() );
+	}
+};
 
-		if (
-			nextProps.primaryAndRecent &&
-			! isEqual( nextProps.recentSiteIds, this.props.recentSiteIds )
-		) {
-			this.requestRecent( nextProps );
-		}
+const requestSingle = siteId => ( dispatch, getState ) => {
+	if ( siteId && ! isRequestingSite( getState(), siteId ) ) {
+		dispatch( requestSite( siteId ) );
 	}
+};
 
-	requestAll( props ) {
-		if ( props.allSites && ! props.requestingSites ) {
-			props.requestSites();
-		}
+const requestPrimary = siteId => ( dispatch, getState ) => {
+	if ( siteId && ! hasAllSitesList( getState() ) && ! isRequestingSite( getState(), siteId ) ) {
+		dispatch( requestSite( siteId ) );
 	}
+};
 
-	requestPrimary( props ) {
-		if ( props.primaryAndRecent && ! props.hasAllSitesList ) {
-			const { primarySiteId, isRequestingPrimarySite } = props;
+const requestRecent = siteIds => ( dispatch, getState ) => {
+	if ( siteIds.length && ! hasAllSitesList( getState() ) ) {
+		const isRequestingSomeSite = siteIds.some( siteId => isRequestingSite( getState(), siteId ) );
```

Comment:
I think this logic -- don't request _any_ of the recent sites if a _single one_ of them is being requested -- is a bit unsound. Let's do the check the same way we do it for single or primary site: don't request the site if it's already being requested, independently from any other site.

Reply:
That was actually my initial implementation, but I went back to the behaviour in the old code for two reasons:
- Depending on details of dispatch ordering (which I'm unfamiliar with), it could result in the same site being requested multiple times
- It seemed dangerous to change the logic

I'm happy to make the change if you don't think those concerns are an issue. It would actually make the code much simpler!

Reply:
I think it's fine and beneficial to change the implementation. Also, now when the `isRequesting` check and `request` execution are done synchronously within a single action thunk, we can be much more sure that duplicate requests won't be sent. It wasn't so certain when the `isRequesting` prop was a connected prop of a React component and the update went through the React lifecycle machinery.

The change will also let us simplify and unify the code for `requestPrimary` and `requestRecent` -- the only difference will be that one works with a single site and the other works with an array.

Also, carefully filtering `primarySiteId` from `recentSiteIds` list is not needed any more -- it will be quickly deduplicated by the `isRequesting` check.

The main reason why `<QuerySites primaryAndRecent />` exists is to populate the most interesting part of the sites list as quickly as possible, and avoid waiting for the full list. For users with hundreds of sites (mostly a12s 😉), this makes a significant difference.

Reply:
Fair enough, that all makes perfect sense! I've simplified the logic for this, using a single array for both primary and recent (letting the dispatch handle the deduplication), and reusing `requestSingle` for each individual request.